fw/cache.c

```diff
@@ -511,9 +511,34 @@ tfw_cache_employ_resp(TfwHttpResp *resp)
 	if ((req->cache_ctl.flags & TFW_HTTP_CC_HDR_AUTHORIZATION)
 	    && !(effective_resp_flags & CC_RESP_AUTHCAN))
 		return false;
-	if (!(effective_resp_flags & CC_RESP_CACHEIT)
-	    && !tfw_cache_status_bydef(resp))
-		return false;
+	if (!(effective_resp_flags & CC_RESP_CACHEIT)) {
+		/*
+		 * According to RFC 9110 9.3.3:
+		 * Responses to POST requests are only cacheable when
+		 * they include explicit freshness information.
+		 */
+		if (req->method == TFW_HTTP_METH_POST)
+			return false;
+		if (!tfw_cache_status_bydef(resp))
+			return false;
+	}
+
+	/*
+	 * According to RFC 9110 9.3.3:
+	 * Responses to POST requests are only cacheable when
+	 * they include a Content-Location header field that
+	 * has the same value as the POST's target URI.
+	 */
+	if (req->method == TFW_HTTP_METH_POST) {
+		TfwStr *h = &resp->h_tbl->tbl[TFW_HTTP_HDR_CONTENT_LOCATION];
+		TfwStr h_val;
+
+		tfw_http_msg_srvhdr_val(h, TFW_HTTP_HDR_CONTENT_LOCATION,
+					&h_val);
+		if (tfw_strcmp(&h_val, &req->uri_path))
```

Comment:
This is sophisticated part and I'm not sure, that we must implement it strictly to RFC, but I must draw your attention. URI in this case might be in two forms absolute-URI and partial-URI, if request had partial-URI, but `content-location` in response has absolute-URI in current implementation Tempesta won't cache such response. However, rfc says that we should transform partial-URI to absolute before comparison:

> If Content-Location is included in a 2xx (Successful) response message and its value refers (after conversion to absolute form)  to a URI that is the same as the target URI

This conversion also not obvious to me, because absolute form of URI has scheme information and there is question what scheme we must use during this conversion? And another question what we must do if both URI have absolute form with different schemes(http vs https)?

Reply:
req->uri_path contains relative URI. Yes, in this simple implementation we rely on the fact that server sends us response with Content-Location with appropriate relative format. If it is not so we don';t cache response. @krizhanovsky should I implement Content-Location parser and comparison as it described in RFC?

Reply:
Perhaps, there's the fourth solution: make this kind of behavior configurable to allow either one of these three strategies. The downside of this is that there may be too many configurable options already which adds to the complexity of Tempesta's use.

On this thought, it might be interesting to collect the configurations from real users/admins (anonymously, of course) to see what options are really used. That may lead to a less complex product. It's like simplicity and minimalism of Apple products versus the magnitude of options of Android systems. :-)

Reply:
For some time ago we researched abosolute URIs in the wild and no legitimate clients actually produce such requests. IIRC mostly such requests are used for forward proxies, but here I could be wrong. In either case we already made some steps in not supporting absolute URLs. I thing we're good here with the current implementation.

I'd only add a good comment what RFC requires, what do we do and why.